### ISSUE TITLE
Crates.io returns mostly 404

### ISSUE THREAD

**Issue by kud1ing** [rust-lang/crates.io#788]
Maybe this is a duplicate of other 404-issues here, i am not sure.

`awesome-rust` uses `awesome_bot` to check links.
Unfortunately many (all?) links to crates.io return 404: https://travis-ci.org/rust-unofficial/awesome-rust

`curl https://crates.io/keywords/cassandra` receives a 404, too.
I am a web noob. Is there somwething that needs to be configured on the server or client?

**Comment by carols10cents** [comment]
I think we are doing something nonstandard, but I'm not sure what. Until we figure this out, a workaround is specifying HTML as the content type you want:

```
curl -v 'https://crates.io/crates/assert_approx_eq' -H 'Accept: text/html'
```

returns a 200 for me. Hope that helps!

**Comment by kud1ing** [comment]
Please give us a heads up when this is resolved.
`crates.io` was whitelisted in https://github.com/rust-unofficial/awesome-rust/pull/310
I'd like to revert this finally.

**Comment by Nemo157** [comment]
Yeah, this is caused by https://github.com/rust-lang/crates.io/blob/a20eb96877d32bff6b5b9ac8ea3dfc497b5351b9/src/dist.rs#L38-L52

Interestingly requesting _any_ URL with `Accept: text/html` will return 200 since the server doesn't know what the valid routes are (e.g. `curl -H 'Accept: text/html' -I https://crates.io/foo/bar`)

I don't think there's any way to have a bot validate URLs on crates.io without executing the javascript and checking if it loads the page successfully. At least, while crates.io does client-side only rendering, if server-side rendering of at least the initial page is ever added that should then know whether the URL being requested is valid or not.

**Comment by locks** [comment]
FYI, you can track server-side rendering efforts at https://github.com/rust-lang/crates.io/pull/819.

**Comment by AaronFriel** [comment]
This is still not fixed.

![image](https://user-images.githubusercontent.com/788800/34024765-a455a552-e111-11e7-9e82-15c780a0e941.png)

**Comment by carols10cents** [comment]
@AaronFriel Yes, that's why this issue is still open.

**Comment by AaronFriel** [comment]
Would a commit altering the `wants_html` test to include wildcards be accepted? I'm just not sure why this has languished and I'm guessing there is something I'm missing.

**Comment by carols10cents** [comment]
Yes, I am not sure of the scope of the solution needed or the effects of various solutions on the frontend and backend-- please give it a try and let me know if you have questions!

**Comment by mark-i-m** [comment]
Any progress on this?

**Comment by luciusmagn** [comment]
Any updates on this?

**Comment by jtgeibel** [comment]
I've recently added PR #1788 which would impact this bug.  This PR would make it so that the behavior is consistent, regardless of if an `Accept: text/html` header was sent or not.  All such requests will now return a status 200 with the Ember index html (as is currently done for browsers and other clients that set the header).

This will in effect invert the issue described here.  Instead of always returning a 404 with a JSON response, the site will now always respond with a status 200 with static HTML.  Instead of false negatives for crates that do exists, there would be false positives for crates that do not exist.

**Comment by 89z** [comment]
Just ran into this. This works:

~~~
curl -I -H Accept:text/html https://crates.io
~~~

but if you remove the header, you get 404. I have never seen this requirement in any server I have ever dealt with, and I write extractors that deal with all kinds of server nonsense. Cant you just look at any of the hundreds of open source Rust server implementations, and see what they are doing?
